### ISSUE TITLE
Unauthenciated google grpc request crash envoy

### ISSUE THREAD

**Issue by fshewl** [envoyproxy/envoy#5127]
**Issue Template**

*Title*: *Unauthenciated google grpc request crash envoy*

*Description*:
If a google grpc request gets rejected wiht "Unauthencicated" error, envoy crashes.

The reason is, envoy compares the grpc_status code with DATA_LOSS (enum = 15), any error code larger than this number would be converted to InvalidCode (enum = -1)
https://github.com/envoyproxy/envoy/blob/97fa8855d115aed81ebe423135b8f023f41f64a7/source/common/grpc/google_async_client_impl.cc#L332

However, notifyRemoteClose uses the grpc_status_code to access an array. -1 index cause it to crash
https://github.com/envoyproxy/envoy/blob/97fa8855d115aed81ebe423135b8f023f41f64a7/source/common/grpc/google_async_client_impl.cc#L183

Stack:
[third_party/envoy/src/source/common/grpc/google_async_client_impl.cc:326] Finish with grpc-status code 16
[2018-11-26 22:36:00.404][000001][debug][grpc] [third_party/envoy/src/source/common/grpc/google_async_client_impl.cc:182] notifyRemoteClose -1 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.

**Comment by fshewl** [comment]
Worth noting that DATA_LOSS is not the max grpc error code anymore.

Unauthenticated has enum 16.

**Comment by fshewl** [comment]
https://github.com/googleapis/googleapis/blob/3cef6c237ba87a81470b5fdf082ed187cbb23ded/google/rpc/code.proto#L103

**Comment by qiannawang** [comment]
@mattklein123  and @htuch, I think the fix is to change the validation condition to use Status::GrpcStatus::MaximumValid instead of grpc::StatusCode::DATA_LOSS in envoy/source/common/grpc/google_async_client_impl.cc. In particular, the validation becomes

const Status::GrpcStatus grpc_status =
        status_.error_code() <= Status::GrpcStatus::MaximumValid
            ? static_cast<Status::GrpcStatus>(status_.error_code())
            : Status::GrpcStatus::InvalidCode;

As @fshewl pointed out, DATA_LOSS is no longer the largest valid code any more. In Status::GrpcStatus, it does provide the number for the valid status. 

Let me know what you think about the proposed fix? If reasonable, my team would like to take this.

**Comment by mattklein123** [comment]
@qiannawang yes that sounds about right. Please make sure you add a test also for a code that does not yet exist and we can discuss further in the fix PR.

**Comment by htuch** [comment]
SGTM, please also update the HTTP <-> gRPC status mapping code in https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/status.cc as well.

**Comment by qiannawang** [comment]
SG. @mattklein123, @htuch can you assign this issue to @fshewl ? I dont have the permission to do so.

**Comment by htuch** [comment]
@qiannawang done, but check out https://github.com/envoyproxy/envoy/blob/master/source/docs/repokitteh.md for next time, we have a bot to help! :)

**Comment by qiannawang** [comment]
Cool, thanks for the pointer!

**Comment by fshewl** [comment]
> SGTM, please also update the HTTP <-> gRPC status mapping code in https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/status.cc as well.

I think it already includes "Unauthenticated" in the map:
https://github.com/envoyproxy/envoy/blob/890f1f13179e74d4b728599da66466acf0338762/source/common/grpc/status.cc#L78
